### ISSUE TITLE
Override compiler version

### ISSUE THREAD

**Issue by vbrandl** [kolloch/crate2nix#45]
How can I override, which rustc version or branch (stable/beta/nightly) to use?

Also it would be cool, if `crate2nix` would check for `rust-toolchain` files which are used by rustup to pin a specific compiler version and or branch for a repository.

**Comment by kolloch** [comment]
You can override the rust compiler by overlays to the used nixpkgs. 

I will investigate in making this easy...

**Comment by vbrandl** [comment]
Would be cool to mention this in the readme with an example.

I tried adding the mozilla overlay but I can't figure out, where do override the used compiler channel

**Comment by andir** [comment]
I am using the following pattern in an overlay:

```nix
self: super: {
   rustNightlyChannel = mozilla.rustChannelOf { date = "..."; channel = "..."; };
   rustNightly = self.rustNightlyChannel.rust;
   myBuildRustCrate = super.buildRustCrate.override { rustc = self.rustNightly; };
   myPkg = (self.callPackage ./Crate.nix { buildRustCrate = myBuildRustCrate; }).rootCrate.build;
}
```
(adopted from my actual, more complicated, version)

**Comment by vbrandl** [comment]
Thank you. I also found this one yesterday but I didn't come around to test it, yet: https://github.com/jD91mZM2/xidlehook/blob/master/default.nix

**Comment by cpcloud** [comment]
@kolloch @vbrandl Can this be closed?

**Comment by danieldk** [comment]
By the way, you do not have to use an overlay, `buildRustCrate` allows you to specify a Rust toolchain through the `rust` argument. See here for an example that builds a crate using nightly:

https://github.com/finalfusion/nix-packages/blob/cf8a4140d2359fb512608bde3c88ba410b5d94fb/pkgs/python-modules/finalfusion/default.nix

**Comment by vbrandl** [comment]
I wasn't able to test yet because I didn't need it anymore (since 1.39.0 landed in the unstable channel, I don't need a nightly compiler for async/await). This can be closed but I think it would be nice to have a documented example in the README that shows how to use the mozilla overlay

**Comment by kolloch** [comment]
@andir @vbrandl @danielk I added documentation for doing this "globally" here:

https://github.com/kolloch/crate2nix#choosing-a-rust-version

I discovered one caveat: I had to use "rust" (without the c suffix) for "rustc" with a recent version of the overlay because otherwise `rustc` wouldn't find the std sources.

Doing this globally is often what you want, I guess. Build all your stuff with the same rust compiler. But the individual approach also has merrit. Maybe you want to add that to the documentation? You would be welcome.

@danielk for making buildRustCrate more directly overridable, see https://github.com/kolloch/crate2nix/issues/77. Tell me what you think! For "pkgs" it is not the right tool, though, since you also want to override it for all dependencies, correct?
